species = 'saccharomyces_cerevisiae'
gtf = 'Saccharomyces_cerevisiae.R64-1-1.80.gtf'
fasta = 'Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa'
kidx = 'Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.kidx'
ensembl = 'ftp://ftp.ensembl.org/pub/release-80'

sids = ['SRR1812768', 'SRR1812760', 'SRR1811834', 'SRR1812754', 'SRR1812759',
        'SRR1812767']

rule all:
    input:
        kidx,
        expand('results/{s}/kallisto/abundance.h5', s = sids)

rule get_transcriptome:
    output:
        expand('../genome/{f}', f = fasta)
    shell:
        'cd ../genome && '
        'wget -O {fasta}.gz {ensembl}/fasta/{species}/dna/{fasta}.gz && '
        'gunzip {fasta}.gz'

rule get_anno:
    input:
        expand('../genome/{f}', f = fasta)
    output:
        expand('../annotation/{g}', g = gtf)
    shell:
        'cd ../annotation && '
        'wget -O {gtf}.gz {ensembl}/gtf/{species}/{gtf}.gz && '
        'gunzip {gtf}.gz && '
	'gffread {gtf} ../genome/{fasta}'

rule index:
    input:
        expand('../genome/{f}', f = fasta)
    output:
        kidx
    shell:
        'kallisto index '
        '-k 21 '
        '-i {output} '
        '../genome/{fasta}'

rule kallisto:
    input:
        'results/{samp}/{samp}_1.fastq.gz',
        'results/{samp}/{samp}_2.fastq.gz',
        kidx
    output:
        'results/{samp}/kallisto',
        'results/{samp}/kallisto/abundance.h5'
    threads: 2
    shell:
        'kallisto '
        'quant '
        '-i {kidx} '
        '--bias '
        '-b 30 '
        '-o {output[0]} '
        '-t {threads} '
        '{input[0]} {input[1]}'

rule fastq_dump:
    output:
        'results/{samp}/{samp}_1.fastq.gz',
        'results/{samp}/{samp}_2.fastq.gz'
    threads: 1
    shell:
        'fastq-dump '
        '--split-files '
	'-O results/{wildcards.samp} '
        '--gzip '
        '{wildcards.samp}'

rule clean:
    shell:
        'rm -rf results/*/kallisto {kidx} {fasta}'
